src/basic/fileio.c

```diff
@@ -259,7 +259,14 @@ int read_full_stream(FILE *f, char **contents, size_t *size) {
                 if (ferror(f))
                         return -errno;
 
-                if (feof(f))
+                /* stdio will only set EOF after attempting to read past the
+                 * end of the file.  In the common regular file case where
+                 * know, allocate, and read the file's exact size, EOF isn't
+                 * found yet.  Do our own check against size here to avoid
+                 * doubling the buffer just for fread() to read nothing and
+                 * find eof on the next iteration.
+                 */
+                if (feof(f) || (st_size > 0 && st_size == l))
```

Comment:
Having thought about this a bit more, and honestly preferring a more general solution which also avoids allocating 2x the space for proc files equally well, I'd like to propose a different approach:

Perform a 1 byte readahead into a local variable when feof() is false, if it's read successfully the next iteration will consume it as if fread() had (adjusting all indices accordingly), if it's not read and we've found EOF then behave as if feof() returned true.

With the FILE * being block buffered, I don't think this would be a significant performance difference, and it allows us to only ever enlarge the buffer when we know there's at least one more byte to be read.

Does that make sense?